GSM SIM fixes

laf0rge opened the pull request:
This branch contains GSM SIM related fixes. With those applied, I can get seveal legacy 2G only MS to recognize the card and authenticate to a network.

laf0rge committed:
GSM SIM: Advertise as 1.8/3/5V capable SIM

If we don't do this, then almost all real-world phones will reject
to use the card, as they can only supply 1.8 or 3V.  Only very early
GSM phones (think of portable suitcase phones, not handset) could
actually provide 5V to the SIM.

Closes #6

laf0rge committed:
GSM READ BINARY: Fix length / bounds check

We must verify that the offset plus the expected/read length doesn't
exceed the end of file.

tomasz-lisowski commented on src/gsm.c:
```diff
@@ -152,18 +152,17 @@ int32_t gsm_select_res(swicc_fs_st const *const fs,
          *  + 1b Clock stop             = 0 (low not preferred)
-         *  + 1b 0 (from GSM 11.12 @todo Look into GSM 11.12)
-         *  + 2b RFU                    = 00
+         *  + 3b Supply voltage ind.    = 3 (1.8V Technology SIM (supports 3V/5V)
```
I set this to `0b000` because per ETSI TS 102 221 (v16.4.0) 6.2.1, if the card supports multiple supply voltage classes, they need to be consecutive, i.e., AB, BC, ABC, but not AC. Setting this to `0b011` would mean classes A and B would appear unsupported? Is there a mention in any of the specs about support for classes A and B when only C is indicated? Technically, ETSI TS 102 221 also extends the ISO classes with D and E. In summary, I don't know what indication to use, in order to support all the supply voltage classes and maximize compatibility.

Missing ')', and I'd prefer if you swapped `3` to binary `011` as the other flags are all listed in binary.

laf0rge replied:
I've fixed the parenthesis and the decimal/binary in my updated branch/PR.

Regarding your reference to ETSI TS 102 221 - I don't really understand why it would apply here. Classic GSM SIM is not an UICC application. So unless TS 51.011 explicitly (directly or indirectly e.g. via 31.101) references ETSI TS 102 221, I don't see why or how it would apply.  Particularly  since TS 51.011 gives its very own  definition of those bits. Also, the value `011` was taken from a real-world T-Mobile SIM card, which works in all the phones I've tested.

tomasz-lisowski replied:
You're absolutely right. I also checked in GSM 11.12, and it seems 3v SIM cards must also support 5v, which should increase compatibility.

tomasz-lisowski commented:
The 2G devices I tested on didn't mind the 5v indication (e.g. Samsung GT-E1150i, and Nokia E72 both worked). Granted, I didn't know what voltages are really applied to the SIM in any of the devices I tested. Also, thank you for testing swSIM with more devices, I didn't have access to many 2G-only devices to test with, and it's likely that more of these types of issues will arise with more use.

As a sidenote, I noticed that SIMtrace didn't work with Samsung i600, but I didn't have time to investigate that issue further. Even with the stock SIMtrace firmware and a card inserted directly into SIMtrace, the device still rejected the card.

laf0rge committed:
GSM SIM: Support UPDATE BINARY command

TS 51.011 references 31.101 which in turn references TS 102 221,
so we can just wrap the UICC code in one more constraint check.

Closes #8

laf0rge commented:
> The 2G devices I tested on didn't mind the 5v indication (e.g. Samsung GT-E1150i, and Nokia E72 both worked). 
> Also, thank you for testing swSIM with more devices, I didn't have access to many 2G-only devices to test with, and it's likely that more of these types of issues will arise with more use.

I probably have more than 50 different types of 2G MS here, but I don't have any plan to do extensive swsim compatibility testing due to time constraints.  I was merely using those devices that I happened to use for the higher-level task that I'm currently working on (GSM CSD support in Osmocom CNI) - and for that I needed a SIM card with 2G XOR authentication, which sadly none of my programmable cards can do, so I resorted to a simulated SIM and hacked XOR-2G into swsim (ugly #ifdef, not worth submitting as-is).  All three 2G MS I tried yesteday (Siemens AC75i, Siemens BGS2E, Motorola C121) gave up with that 5V-only coding.

> As a sidenote, I noticed that SIMtrace didn't work with Samsung i600, but I didn't have time to investigate that issue further. 

sorry to hear.

> Even with the stock SIMtrace firmware and a card inserted directly into SIMtrace, the device still rejected the card.

Now that sounds relatively unlikely, so I would almost assume that it is some kind of contact problem.  I'm happy to send you a combination of https://shop.sysmocom.de/Full-size-SIM-card-to-FPC-adapter/fullsize-sim-fpc and https://shop.sysmocom.de/Professional-SIM-card-adapter-plug-in-micro-nano-SIM-to-full-size/sim-adapter-pcb free of charge, so you could test with just "extending" the SIM slot to the card without any active electronics involved.  IF that already fails, it's a mechanical/contact problem, or potentially an EMI issue.

tomasz-lisowski commented:
> Now that sounds relatively unlikely, so I would almost assume that it is some kind of contact problem. I'm happy to send you a combination of https://shop.sysmocom.de/Full-size-SIM-card-to-FPC-adapter/fullsize-sim-fpc and https://shop.sysmocom.de/Professional-SIM-card-adapter-plug-in-micro-nano-SIM-to-full-size/sim-adapter-pcb free of charge, so you could test with just "extending" the SIM slot to the card without any active electronics involved. IF that already fails, it's a mechanical/contact problem, or potentially an EMI issue.

Indeed, I could try that. I also need to get myself a new SIMtrace as the one I used while developing swSIM was borrowed :).
